compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1234,18 +1235,26 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
           // if expected parameter type(s) are wildcards, approximate from below.
           // if expected result type is a wildcard, approximate from above.
           // this can type the greatest set of admissible closures.
-          (pt1.argTypesLo.init, typeTree(interpolateWildcards(pt1.argTypesHi.last)))
+          // However, we still keep the information on whether expected parameter types were 
+          // wildcards, in case of types inferred from target being more specific
+
+          val fromWildcards = pt1.argInfos.init.map{
+            case bounds @ TypeBounds(nt, at) if nt == defn.NothingType && at == defn.AnyType => true
```

Comment:
Type comparison via `==` or `!=` is not recommended. One can compare types with `eq`, or pattern match the structure and compare symbols. Or else use `=:=` which is semantic equality, equivalent to mutual subtyping, but this one is expensive. 

In this specific case, there are detectors for the types your are interested in: `isExactlyNothing` and `isAny`.